tests/unit/test_spmd.c

```diff
@@ -0,0 +1,254 @@
+
+#define TEST_MAX_GATHER_BLOCK_SIZE 32
+
+/* The actual tests are here. */
+int run_spmd_tests(MPI_Comm test_comm)
+{
+    int my_rank;  /* 0-based rank in test_comm. */
+    int ntasks;   /* Number of tasks in test_comm. */
+    int num_elem; /* Number of elements in buffers. */
+    int *sbuf;    /* The send buffer. */
+    int *rbuf;    /* The receive buffer. */
+    struct timeval t1, t2; /* For timing. */
+
+    /* Used for testing pio_swapm. */
+    int *sendcounts;
+    int *recvcounts;
+    int *rdispls;
+    int *sdispls;
+    MPI_Datatype *sendtypes;
+    MPI_Datatype *recvtypes;
+
+    int mpiret;   /* Return value from MPI calls. */
+    int ret;      /* Return value. */
+
+    /* Learn rank and size. */
+    if ((mpiret = MPI_Comm_size(test_comm, &ntasks)))
+        MPIERR(mpiret);
+    if ((mpiret = MPI_Comm_rank(test_comm, &my_rank)))
+        MPIERR(mpiret);
+
+    /* Determine size of buffers. */
+    num_elem = ntasks * ntasks;
+
+    /* Allocatte the buffers. */
+    if (!(sbuf = malloc(num_elem * sizeof(int))))
+        return PIO_ENOMEM;
+    if (!(rbuf = malloc(num_elem * sizeof(int))))
+        return PIO_ENOMEM;
+
+    /* Test pio_swapm Create and load the arguments to alltoallv */
+
+    /* Allocate memory for the arrays that are agruments to the
+     * alltoallv. */
+    if (!(sendcounts = malloc(ntasks * sizeof(int))))
+        return PIO_ENOMEM;
+    if (!(recvcounts = malloc(ntasks * sizeof(int))))
+        return PIO_ENOMEM;
+    if (!(rdispls = malloc(ntasks * sizeof(int))))
+        return PIO_ENOMEM;
+    if (!(sdispls = malloc(ntasks * sizeof(int))))
+        return PIO_ENOMEM;
+    if (!(sendtypes = malloc(ntasks * sizeof(MPI_Datatype))))
+        return PIO_ENOMEM;
+    if (!(recvtypes = malloc(ntasks * sizeof(MPI_Datatype))))
+        return PIO_ENOMEM;
+
+    /* Initialize the arrays. */
+    for (int i = 0; i < ntasks; i++)
+    {
+        sendcounts[i] = i + 1;
+        recvcounts[i] = my_rank + 1;
+        rdispls[i] = i * (my_rank + 1) * sizeof(int);
+        sdispls[i] = (((i+1) * (i))/2) * sizeof(int);
+        sendtypes[i] = recvtypes[i] = MPI_INT;
+    }
+
+    /* Free memory. */
+    free(sendcounts);
+    free(recvcounts);
+    free(rdispls);
+    free(sdispls);
+    free(sendtypes);
+    free(recvtypes);
+
+
+    //    for (int msg_cnt=4; msg_cnt<size; msg_cnt*=2){
+    //   if (rank==0) printf("message count %d\n",msg_cnt);
+    int msg_cnt = 0;
+    for (int itest = 0; itest < 5; itest++)
+    {
+        bool hs = false;
+        bool isend = false;
+
+        /* Load up the buffers */
+        for (int i = 0; i < num_elem; i++)
+        {
+            sbuf[i] = i + 100 * my_rank;
+            rbuf[i] = -i;
+        }
+        MPI_Barrier(test_comm);
+
+        if (!my_rank)
+        {
+            printf("Start itest %d\n", itest);
+            gettimeofday(&t1, NULL);
+        }
+
+        if (itest == 0)
+            ret = pio_swapm(sbuf, sendcounts, sdispls, sendtypes, rbuf, recvcounts,
```

Comment:
I agree! I wanted to get these changes merged, then will move on with this test.

Now that I have added your very helpful comments I will do the merge. Thanks!! ;-)